src/main/cpp/src/shuffle_assemble.cu

```diff
@@ -0,0 +1,1771 @@
+          partitions + partition_offsets[partition_index] + sizeof(partition_header));
+        auto const col_index = i / num_partitions;
+
+        return has_validity_buf[col_index / 32] & (1 << (col_index % 32)) ? 1 : 0;
+      }));
+  thrust::reduce_by_key(rmm::exec_policy_nosync(stream, temp_mr),
+                        column_keys,
+                        column_keys + num_column_instances,
+                        has_validity_values,
+                        thrust::make_discard_iterator(),
+                        assemble_column_info_has_validity_output_iter{column_info.begin()},
+                        thrust::equal_to<size_type>{},
+                        thrust::logical_or<bool>{});
+
+  // compute everything else except row count (which will be done later after we have computed
+  // column instance information)
+  auto iter = thrust::make_counting_iterator(0);
+  thrust::for_each(rmm::exec_policy_nosync(stream, temp_mr),
+                   iter,
+                   iter + num_columns,
+                   [column_info     = column_info.begin(),
+                    global_metadata = global_metadata.begin()] __device__(size_type col_index) {
+                     auto const& metadata = global_metadata[col_index];
+                     auto& cinfo          = column_info[col_index];
+
+                     cinfo.type         = metadata.type;
+                     cinfo.valid_count  = 0;
+                     cinfo.num_children = metadata.num_children();
+                   });
+
+  // generate per-column-instance data ------------------------------------------------------
+
+  // has-validity, type, # of children, row count for non-offset child columns
+  thrust::for_each(
+    rmm::exec_policy_nosync(stream, temp_mr),
+    iter,
+    iter + num_column_instances,
+    [column_instance_info = column_instance_info.begin(),
+     global_metadata      = global_metadata.begin(),
+     partitions           = partitions.data(),
+     partition_offsets    = partition_offsets.begin(),
+     num_columns,
+     per_partition_metadata_size] __device__(size_type i) {
+      auto const partition_index    = i / num_columns;
+      auto const col_index          = i % num_columns;
+      auto const col_instance_index = (partition_index * num_columns) + col_index;
+
+      auto const& metadata = global_metadata[col_index];
+      auto& cinstance_info = column_instance_info[col_instance_index];
+
+      uint8_t const* const buf_start =
+        reinterpret_cast<uint8_t const*>(partitions + partition_offsets[partition_index]);
+      partition_header const* const pheader = reinterpret_cast<partition_header const*>(buf_start);
+
+      bitmask_type const* const has_validity_buf =
+        reinterpret_cast<bitmask_type const*>(buf_start + sizeof(partition_header));
+      cinstance_info.has_validity =
+        has_validity_buf[col_index / 32] & (1 << (col_index % 32)) ? 1 : 0;
```

Comment:
Here is that calculation again. Should we bump this out into a function?

Reply:
Done.